File: math/primenumbers/202-happy-number/happy-number.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
//    let str=(String)(n)
//    let sum=0
//     while((Number)(str)!=1){
//      sum=0
//     for(let i=0;i<str.length;i++){
//         let m=(Number)(str[i])
//         sum+=m*m
//     }
//     str=sum.toString()
//     }

    // return true

    function add(num,sum=0){//19
    
   while(num>0){
      let dig=num%10//9
     sum+=dig*dig//81

     num=Math.floor(num/10)
   }

   return sum
    }
   
   let s=n
   let f=add(n)
  
  while(f!=s && f!=1){
    s=add(s)
    f=add(add(f))

  }

    return f==true
};

File: math/primenumbers/202-happy-number/happy-number.java
class Solution {
     
     private static boolean sum(HashSet<Integer> set,int n){

        if(n==1){
            return true;
        }
        else  if(set.contains(n)){
            return false;
        }
         set.add(n);

        int sum=0;

        while(n!=0){
            int l=n%10;
            int sq=l*l;
            sum+=sq;
            n=n/10;
        }

       

       
         return sum(set,sum);
     }

    public boolean isHappy(int n) {

                HashSet<Integer> set=new HashSet<>();
        return sum(set,n);

        
        
        
    }
}